prediction-applier/src/main/java/org/alfresco/hxi_connector/prediction_applier/hx_insight/HxPredictionReceiver.java

```diff
@@ -0,0 +1,135 @@
+    private static final String IS_PREDICTION_PROCESSING_PENDING_KEY = "is-prediction-processing-pending";
+    private static final String HAS_NEXT_PAGE_KEY = "has-next-page";
+    private static final String PREDICTIONS_BATCH_KEY = "predictions-batch";
+
+    private final PredictionListenerConfig config;
+
+    @Override
+    public void configure()
+    {
+        // @formatter:off
+        JacksonDataFormat predictionsBatchDataFormat = new ListJacksonDataFormat(Prediction.class);
+        JacksonDataFormat predictionDataFormat = new JacksonDataFormat(Prediction.class);
+
+        from(config.predictionProcessorTriggerEndpoint())
+                .routeId(PREDICTION_PROCESSOR_TRIGGER_ROUTE_ID)
+                .choice()
+                .when(this::isProcessingPending)
+                    .log(DEBUG, log, "Prediction processing is pending, no need to trigger it")
+                .otherwise()
+                    .log(DEBUG, log, "Triggering prediction processing")
+                    .to(PREDICTION_PROCESSOR);
+
+        from(PREDICTION_PROCESSOR)
+                .routeId(PREDICTION_PROCESSOR_ROUTE_ID)
+                .process(setIsProcessingPending(true))
+                .loopDoWhile(this::hasNextPage)
+                    .log(DEBUG, log, "Fetching predictions")
+                    .to(config.predictionsSourceEndpoint())
+                    .log(DEBUG, log, "Sending predictions to internal buffer: ${body}")
+                    .unmarshal(predictionsBatchDataFormat)
+                    .process(this::savePredictionsBatch)
+                    .loopDoWhile(this::predictionsBatchNotEmpty)
+                        .process(this::setPredictionToSend)
+                        .marshal(predictionDataFormat)
+                        .log(TRACE, log, "Sending prediction to internal buffer: ${body}")
+                        .to(config.internalPredictionsBufferEndpoint())
+                    .end()
+                .end()
+                .log(DEBUG, log, "Finished processing predictions")
+                .process(setIsProcessingPending(false));
+        // @formatter:on
+    }
+
+    private boolean isProcessingPending(Exchange exchange)
+    {
+        return Objects.requireNonNullElse(
+                getContext().getRegistry().lookupByNameAndType(IS_PREDICTION_PROCESSING_PENDING_KEY, Boolean.class),
+                false);
+    }
+
+    private Processor setIsProcessingPending(boolean isProcessingPending)
+    {
+        return exchange -> getContext().getRegistry().bind(IS_PREDICTION_PROCESSING_PENDING_KEY, isProcessingPending);
+    }
+
+    private boolean hasNextPage(Exchange exchange)
+    {
+        return Objects.requireNonNullElse(exchange.getVariable(HAS_NEXT_PAGE_KEY, Boolean.class), true);
+    }
+
+    private void savePredictionsBatch(Exchange exchange)
+    {
+        LinkedList<Prediction> predictionsBatch = new LinkedList(exchange.getIn().getBody(List.class));
```

Comment:
Why do you need to save the batch? I assume it's still in exchange's body

Reply:
Maybe it would be better to unmarshal the stuff directly to `LinkedList`?

Reply:
unfortunately it is not easy with camel :/ I would leave it as it is as it costs us almost nothing to create new list